### PR TITLE
feat(extras): add typescript vscode extra

### PR DESCRIPTION
**Pull request by Demianeen**
Hi! Spend a month tweaking my typescript configuration to integrate it with vscode. 

How is it different from typescript lang extra?

1. It uses Vtsls instead of Tsserver which has better memory usage (at least from my testing)
2. It uses the cmd .vscode/launch.json for dap configuration. Which allows to define dap configuration per project and use them alongside VSCode colleagues)

I am sure it could be a bit cleaner and better, so maybe you or somebody else will change it a bit)

### CODE REVIEW

**Comment by rubiin** [comment]
I prefer typescript-tools instead of tsserver. Tsserver is not suitable for big projects for the same reason

**Review by rubiin — APPROVED** [review]

File: lua/lazyvim/plugins/extras/lang/typescript-vscode.lua [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+    -- we need workspace/didRename for renaming to work with vtsls
+    {
+      "antosha417/nvim-lsp-file-operations",
+      dependencies = {
+        "nvim-lua/plenary.nvim",
+        "nvim-neo-tree/neo-tree.nvim",
+      },
+      opts = {},
+    },
+  },
```
[the remark this answers is not in the document]
Comment (Demianeen) [reply lua/lazyvim/plugins/extras/lang/typescript-vscode.lua:262]: Yeah, I tried. Vtsls requires `workspace/didRenameFiles` lsp notification which is not implemented in neovim. I tried with lazyvim, but it wasn't working as well. So I assume lazyvim hasn't implemented `workspace/didRenameFiles` notification handler

File: lua/lazyvim/plugins/extras/lang/typescript-vscode.lua [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+        },
+      }
+    end,
+    dependencies = {
+      {
+        "yioneko/nvim-vtsls",
+        handlers = {},
+      },
+    },
+  },
```
Comment (dpetka2001) [inline lua/lazyvim/plugins/extras/lang/typescript-vscode.lua:141]:
```suggestion
  },
    {
    "folke/which-key.nvim",
    opts = {
      defaults = {
        ["<leader>ct"] = { name = "+Project", ["🚫"] = "which_key_ignore" },
        ["<leader>cL"] = { name = "+Log", ["🚫"] = "which_key_ignore" },
      },
    },
  },
```
If you would like, you can include that to show prefixes in `which-key` for the 2 corresponding keymaps.

**Comment by Demianeen** [comment]
Are there any news on will it be merged? It is my first PR that takes so long to merge. Is everything alright with it?

**Comment by dpetka2001** [comment]
The maintainer has updated the pinned issue that he will be out until 28 April and be back next week. None other than the maintainer has merge rights, so please be patient until he's back.

File: lua/lazyvim/plugins/extras/lang/typescript-vscode.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (rstacruz) [reply lua/lazyvim/plugins/extras/lang/typescript-vscode.lua:271]: I'm using tsserver, but I've found that nvim-lsp-file-operations makes renaming references after renaming files work on lazyvim. (Anecdotal, haven't really looked too closely)

[the remark this answers is not in the document]
Comment (folke) [reply lua/lazyvim/plugins/extras/lang/typescript-vscode.lua:271]: LazyVim did support lsp renames (without an additional plugin), but there was a bug in my implementation. I didnt' closely follow the spec. Should all be fixed now.

I've also add a `<leader>cR` to rename the current buffer in a correct way.

lsp renames in LazyVim are enabled for neo-tree, mini.files.

Apart from that, I've been migrating the regular typescript extra to vtsls. Will make a new release soon.

I've tested both typescript-tools.nvim and vtsls and the latter seems the better replacement.

**Comment by folke** [comment]
I'm going to push a bunch of changes to the Typescript and Vue extra so that both use vtsls.

I'll leave this PR open for now, for all the DAP stuff. not sure what part of that is needed or would be better than what we currently have in the typescript extra.

**Comment by folke** [comment]
Done, see: https://github.com/LazyVim/LazyVim/commit/fba06ce9f522b91be8a342f9c028948c2733132d

**Comment by folke** [comment]
I did not include all the keymaps from this PR, since you can always do `<leader>cA` to get all the source actions for any lsp server. For vtsls this includes:

image

**Comment by Demianeen** [comment]
> I'm going to push a bunch of changes to the Typescript and Vue extra so that both use vtsls.
> 
> I'll leave this PR open for now, for all the DAP stuff. not sure what part of that is needed or would be better than what we currently have in the typescript extra.

Sure, I did a dap configuration following this [guide](https://miguelcrespo.co/posts/debugging-javascript-applications-with-neovim)

**Comment by folke** [comment]
Closing since we now use vdtsls in the typescript extra.
Thank you anyway!

**Comment by Demianeen** [comment]
> Closing since we now use vdtsls in the typescript extra. Thank you anyway!

Sure @folke! Should we add dap to vtsls? As for now, it is very simple, and it can be a pain to setup

**Comment by Demianeen** [comment]
> I'm honestly not using dap for typescript often.

What do you usually use for debugging?

**Pull request closed without merging**